In [1]:
import numpy as np
import matplotlib.pyplot as plt

from sklearn import datasets
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GridSearchCV

from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier

In [2]:
digits = datasets.load_digits()
X, y = digits.data, digits.target
#Для логичтической регрессии будем использовать отшкалированные данные
X_scaled = StandardScaler().fit_transform(X)

# Logistic Regression

In [3]:
%%time
parameters = {'C': [0.1, 0.5, 1, 5, 10], 'penalty': ['l1', 'l2']}
lr_grid_search = GridSearchCV(LogisticRegression(random_state=42), parameters, cv=10, scoring='accuracy', n_jobs=-1)
lr_grid_search.fit(X_scaled, y)

Wall time: 19.2 s


In [4]:
print("Best parameters: ", lr_grid_search.best_params_)
print("Accuracy: ", lr_grid_search.best_score_)

Best parameters:  {'C': 0.5, 'penalty': 'l1'}
Accuracy:  0.942682248191


# Random Forest

In [5]:
%%time
parameters = {'n_estimators': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 20, 30, 40, 50, 100, 500, 1000], 'criterion': ['gini', 'entropy']}
rf_grid_search = GridSearchCV(RandomForestClassifier(random_state=42), parameters, cv=10, scoring='accuracy', n_jobs=-1)
rf_grid_search.fit(X, y)

Wall time: 33.1 s


In [6]:
print("Best parameters: ", rf_grid_search.best_params_)
print("Accuracy: ", rf_grid_search.best_score_)

Best parameters:  {'criterion': 'gini', 'n_estimators': 40}
Accuracy:  0.954924874791


# k-Nearest Neighbors

In [7]:
%%time
parameters = {'n_neighbors': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 'weights': ['uniform', 'distance'], 'p' : [1, 2, 3]}
knn_grid_search = GridSearchCV(KNeighborsClassifier(), parameters, cv=10, scoring='accuracy', n_jobs=-1)
knn_grid_search.fit(X, y)

Wall time: 2min 4s


In [8]:
print("Best parameters: ", knn_grid_search.best_params_)
print("Accuracy: ", knn_grid_search.best_score_)

Best parameters:  {'n_neighbors': 3, 'p': 3, 'weights': 'uniform'}
Accuracy:  0.979410127991


# Вывод

kNN с параметрами {'n_neighbors': 3, 'p': 3, 'weights': 'uniform'} показал лучший результат по метрике accuracy = 0.979